File: sdks/go/providers/hume/protocol.go
package hume

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"math"
	"strconv"
	"strings"

	out "github.com/speechswitch/client/sdks/go/generated/hume_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

func packet(data []byte, metadata bool) (out.SynthesisItem, error) {
	if !runtime.ValidUTF8JSON(data) {
		return nil, errors.New("Hume returned invalid JSON")
	}
	var value map[string]any
	if json.Unmarshal(data, &value) != nil || value == nil {
		return nil, errors.New("Hume returned an invalid event")
	}
	nativeError, isError := value["error"].(string)
	if isError || value["type"] == "error" {
		message := "Hume synthesis failed"
		if isError {
			message = nativeError
		}
		if v, ok := value["message"].(string); ok {
			message = v
		}
		code := runtime.Optional[string]{}
		if v, ok := value["code"].(string); ok {
			code = runtime.Some(v)
		}
		return nil, &Error{Message: message, Code: code}
	}
	generation, gOK := value["generation_id"].(string)
	request, rOK := value["request_id"].(string)
	snippet, sOK := value["snippet_id"].(string)
	if !gOK || !rOK || !sOK {
		return nil, errors.New("Hume returned invalid correlation identifiers")
	}
	envelope := out.HumeEnvelope{GenerationId: generation, RequestId: request, CorrelationId: snippet, Timestamps: []out.Timestamp{}}
	switch value["type"] {
	case "audio":
		audio, aOK := value["audio"].(string)
		_, aPresent := value["audio"]
		index, iOK := value["chunk_index"].(float64)
		last, lOK := value["is_last_chunk"].(bool)
		utterance, uOK := value["utterance_index"].(float64)
		if (metadata && !aOK) || (aPresent && !aOK) || !iOK || index < 0 || index > 9007199254740991 || math.Trunc(index) != index || !lOK || (value["utterance_index"] != nil && (!uOK || utterance < 0 || utterance > 9007199254740991 || math.Trunc(utterance) != utterance)) {
			return nil, errors.New("Hume returned an invalid audio event")
		}
		// JSON metadata in binary mode is not another copy of the audio.
		if !metadata {
			return nil, nil
		}
		decoded, err := base64.StdEncoding.Strict().DecodeString(audio)
		if err != nil || strings.ContainsAny(audio, "\r\n") {
			return nil, errors.New("Hume returned invalid base64 audio")
		}
		envelope.Audio = runtime.Some(decoded)
		envelope.ChunkIndex = runtime.Some(index)
		var flag out.HumeEnvelopeIsLastChunk = out.HumeEnvelopeIsLastChunkAsFalse{}
		if last {
			flag = out.HumeEnvelopeIsLastChunkAsTrue{}
		}
		envelope.IsLastChunk = runtime.Some(flag)
		if uOK {
			envelope.InputGroupId = runtime.Some(strconv.FormatInt(int64(utterance), 10))
		}
	case "timestamp":
		mark, _ := value["timestamp"].(map[string]any)
		time, _ := mark["time"].(map[string]any)
		text, tOK := mark["text"].(string)
		kind, _ := mark["type"].(string)
		start, sOK := time["begin"].(float64)
		end, eOK := time["end"].(float64)
		if !tOK || (kind != "word" && kind != "phoneme") || !sOK || !eOK || start < 0 || end < start || end > 9007199254740991 || math.Trunc(start) != start || math.Trunc(end) != end {
			return nil, errors.New("Hume returned an invalid timestamp")
		}
		if !metadata {
			return nil, nil
		}
		var literal out.TimestampKind = out.TimestampKindAsWord{}
		if kind == "phoneme" {
			literal = out.TimestampKindAsPhoneme{}
		}
		envelope.Timestamps = []out.Timestamp{{Kind: literal, Value: text, StartTimeMs: start, EndTimeMs: runtime.Some(end)}}
	default:
		return nil, errors.New("Hume returned an invalid event")
	}
	return out.SynthesisItemAsTimeline{Value: envelope}, nil
}

File: sdks/rust/src/providers/hume/protocol.rs
use crate::{base64, generated::hume_output::*, http::TransportError, json::Raw};

#[derive(Debug, PartialEq, Eq)]
pub struct Error {
    pub message: String,
    pub status: Option<u16>,
    pub code: Option<String>,
}
impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}
impl std::error::Error for Error {}
pub(super) fn failure(message: &str) -> TransportError {
    Box::new(std::io::Error::new(
        std::io::ErrorKind::InvalidData,
        message,
    ))
}

pub(super) fn packet(data: &str, metadata: bool) -> Result<Option<SynthesisItem>, TransportError> {
    let raw = Raw::parse_exact(data).map_err(|_| failure("Hume returned invalid JSON"))?;
    let fields = raw
        .object()
        .map_err(|_| failure("Hume returned an invalid event"))?;
    let native_error = fields.get("error").and_then(|v| v.string().ok());
    let kind = fields.get("type").and_then(|v| v.string().ok());
    if native_error.is_some() || kind.as_deref() == Some("error") {
        return Err(Box::new(Error {
            message: fields
                .get("message")
                .and_then(|v| v.string().ok())
                .or(native_error)
                .unwrap_or_else(|| "Hume synthesis failed".into()),
            status: None,
            code: fields.get("code").and_then(|v| v.string().ok()),
        }));
    }
    let mut ids = Vec::new();
    for name in ["generation_id", "request_id", "snippet_id"] {
        ids.push(
            fields
                .get(name)
                .and_then(|v| v.string().ok())
                .ok_or_else(|| failure("Hume returned invalid correlation identifiers"))?,
        );
    }
    let mut e = HumeEnvelope {
        correlation: HumeEnvelopeCorrelation,
        generation_id: ids.remove(0),
        request_id: ids.remove(0),
        correlation_id: ids.remove(0),
        audio: None,
        chunk_index: None,
        is_last_chunk: None,
        input_group_id: None,
        timestamps: vec![],
    };
    match kind.as_deref() {
        Some("audio") => {
            let invalid = || failure("Hume returned an invalid audio event");
            let audio = fields
                .get("audio")
                .map(|v| v.string().map_err(|_| invalid()))
                .transpose()?;
            if metadata && audio.is_none() {
                return Err(invalid());
            }
            let index = fields
                .get("chunk_index")
                .and_then(|v| v.number().ok())
                .ok_or_else(invalid)?;
            let last = fields
                .get("is_last_chunk")
                .and_then(|v| v.boolean().ok())
                .ok_or_else(invalid)?;
            let utterance = fields
                .get("utterance_index")
                .filter(|v| !v.is_null())
                .map(|v| v.number().map_err(|_| invalid()))
                .transpose()?;
            for n in std::iter::once(index).chain(utterance) {
                if !n.is_finite() || n.fract() != 0.0 || !(0.0..=9007199254740991.0).contains(&n) {
                    return Err(invalid());
                }
            }
            // Binary mode's JSON metadata is not a second audio emission.
            if !metadata {
                return Ok(None);
            }
            e.audio = Some(
                base64::decode(&audio.unwrap())
                    .ok_or_else(|| failure("Hume returned invalid base64 audio"))?,
            );
            e.chunk_index = Some(index);
            e.is_last_chunk = Some(if last {
                HumeEnvelopeIsLastChunk::True(HumeEnvelopeIsLastChunkTrue)
            } else {
                HumeEnvelopeIsLastChunk::False(HumeEnvelopeIsLastChunkFalse)
            });
            e.input_group_id = utterance.map(|n| (n as u64).to_string());
        }
        Some("timestamp") => {
            let invalid = || failure("Hume returned an invalid timestamp");
            let mark = fields
                .get("timestamp")
                .and_then(|v| v.object().ok())
                .ok_or_else(invalid)?;
            let time = mark
                .get("time")
                .and_then(|v| v.object().ok())
                .ok_or_else(invalid)?;
            let text = mark
                .get("text")
                .and_then(|v| v.string().ok())
                .ok_or_else(invalid)?;
            let kind = match mark.get("type").and_then(|v| v.string().ok()).as_deref() {
                Some("word") => TimestampKind::Word(TimestampKindWord),
                Some("phoneme") => TimestampKind::Phoneme(TimestampKindPhoneme),
                _ => return Err(invalid()),
            };
            let start = time
                .get("begin")
                .and_then(|v| v.number().ok())
                .ok_or_else(invalid)?;
            let end = time
                .get("end")
                .and_then(|v| v.number().ok())
                .ok_or_else(invalid)?;
            if !start.is_finite()
                || !end.is_finite()
                || start < 0.0
                || end < start
                || end > 9007199254740991.0
                || start.fract() != 0.0
                || end.fract() != 0.0
            {
                return Err(invalid());
            }
            if !metadata {
                return Ok(None);
            }
            e.timestamps.push(Timestamp {
                kind,
                value: text,
                start_time_ms: start,
                end_time_ms: Some(end),
                source: None,
            });
        }
        _ => return Err(failure("Hume returned an invalid event")),
    }
    Ok(Some(SynthesisItem::Timeline(e)))
}
